framework/master-key-provider-keyring.md

```diff
@@ -0,0 +1,125 @@
+A [master key provider](master-key-provider-interface.md) that can decrypt data keys using available [master keys](master-key-interface.md)
+ as well as provide [master keys](master-key-interface.md) for use in encryption.
+
+## Operation
+
+### OnEncrypt
+
+This keyring MUST acquire a list of [master keys](master-key-interface.md) by calling the master key provider's [Get Master Keys for Encryption](master-key-provider-interface.md#get-master-keys-for-encryption)
+using the [encryption context](structures.md#encryption-context) from the [encryption materials](structures.md#encryption-materials) as input. If the 
+master key provider does not provide any master keys, OnEncrypt MUST fail.
+
+If the input [encryption materials](structures.md#encryption-materials) do not contain a [plaintext data key](structures.md#plaintext-data-key), OnEncrypt MUST
+attempt to generate a new plaintext data key and encrypt that data key by calling [Generate Data Key](master-key-interface.md#generate-data-key)
+on the first (primary) [master key](master-key-interface.md) provided by the master key provider.
+
+If the call to Generate Data Key does not succeed, OnEncrypt MUST NOT modify the encryption materials and MUST fail.
+
+If the call succeeds, OnEncrypt MUST do the following with the response from [Generate Data Key](master-key-interface.md#generate-data-key)
+- set the plaintext data key on the [encryption materials](structures.md#encryption-materials)
+- append a new [encrypted data key](structures.md#encrypted-data-key) to the encrypted data key list
+  in the [encryption materials](structures.md#encryption-materials), constructed as follows:
+  - the [ciphertext](structures.md#ciphertext) is the response encrypted data key.
+  - the [key provider id](structures.md#key-provider-id) is the provider id of the primary master key.
+  - the [key provider information](structures.md#key-provider-information) is the key id of the primary master key.
+- append a new record to the [keyring trace](structures.md#keyring-trace) 
+  in the input [encryption materials](structures.md#encryption-materials), constructed as follows:
+  - The string field KeyNamespace contains the provider id of the primary master key.
+  - The string field KeyName contains the key id of the primary master key.
+  - The [flags](structures.md#flags) field of this record includes the following flags:
+    - [GENERATED DATA KEY](structures.md#flags)
+    - [ENCRYPTED DATA KEY](structures.md#flags)
+    - [SIGNED ENCRYPTION CONTEXT](structures.md#flags) (if the master key supports encryption context signing)
+
+Next, for each [master key](master-key-interface.md) provided by the master key provider's 
+[Get Master Keys for Encryption](master-key-provider-interface.md#get-master-keys-for-encryption), the keyring MUST call
+ [EncryptDataKey](master-key-interface.md#encrypt-data-key). If the primary master key was used to generate a data key,
+ that master key MUST NOT be used again for encryption.
```

Comment:
I *think* this should be a SHOULD not a MUST. I'm open to being wrong though.

```suggestion
 that master key SHOULD NOT be used again for encryption.
```

Reply:
I would prefer to default to the more strict MUST NOT, I can't think of why we would want the master key to be used twice and SHOULD NOT leaves the door open for that.